### PR TITLE
Billing models

### PR DESCRIPTION
**Pull request by dylanlott**
Initial pull request for billing models. 

I removed testing references, and included only the model changes and new GraphQL updates. 

I can add testing changes back in if wanted, but I think those would be fitting on a separate branch.

### CODE REVIEW

**Comment by braydonf** [comment]
Notes from what were discussed at noon w/ @bryanchriswhite:

- Seperate PaymentProcess schema out of User and to have a reference to user.
- GraphQL dependency and code is not needed here, as it's duplicated in https://github.com/Storj/billing

Also, in reference to the testing. I didn't mean to remove ALL the tests from this code, but that there were tests included that was not related to these changes, and we could merge those independently.

**Comment by braydonf** [comment]
There are some merge conflicts

File: lib/models/mirror.js [hunk 1]
```diff
@@ -3,6 +3,7 @@
 const errors = require('storj-service-error-types');
 const mongoose = require('mongoose');
 const SchemaOptions = require('../options');
+const _ = require('lodash');
```
Comment (braydonf) [inline lib/models/mirror.js:6]: We probably do not need to bring in lodash specifically for `_.isEmpty`, and lodash is not included in package.json, so this will likely fail.

File: lib/models/mirror.js [hunk 2]
```diff
@@ -38,13 +43,27 @@ Mirror.set('toObject', {
...
   var mirror = new self({
     shardHash: contract.data_hash,
     contact: contact.nodeID,
     contract: contract
   });
 
-  mirror.save(callback);
+  mirror.save(function(err, mirror) {
+    if (err) {
+      return callback(new Error('Cannot save mirror', err));
```
Comment (braydonf) [inline lib/models/mirror.js:63]: Why was this added?

File: lib/models/mirror.js (hunk 1 above)
Comment (barbaraliau) [reply lib/models/mirror.js:6]: Hmm..lodash was there when I wrote the test which is why I used `_.isEmpty`...guess it must have been put it from another merge somewhere. Test can be rewritten to not use lodash.

File: package.json [hunk 3]
```diff
@@ -28,16 +28,17 @@
   "homepage": "https://github.com/Storj/service-storage-models#readme",
   "dependencies": {
     "elliptic": "^6.3.2",
+    "graphql": "^0.8.2",
```
Comment (braydonf) [inline package.json:32]: Not needed here.

File: lib/models/credit.js [hunk 4]
```diff
@@ -5,18 +5,24 @@ const SchemaOptions = require('../options');
 const constants = require('../constants');
 const CREDIT_TYPES = constants.CREDIT_TYPES;
 const PAYMENT_PROCESSORS = constants.PAYMENT_PROCESSORS;
+const PROMO_CODES = constants.PROMO_CODES;
 
-// TODO: standardize property names to lowerCamelCase
+// NB: Check this shit out. Y no pass in earlier?
```
Comment (barbaraliau) [inline lib/models/credit.js:10]: Delete this comment

File: lib/models/user.js [hunk 5]
```diff
@@ -89,14 +186,29 @@ User.set('toObject', {
     delete ret.deactivator;
     delete ret.resetter;
     delete ret.pendingHashPass;
-    delete ret.bytesUploaded;
+    // delete ret.paymentProcessors;
```
Comment (braydonf) [inline lib/models/user.js]: I think these may have been already uncommented before, these should probably not be on the toObject for user?

**Comment by braydonf** [comment]
In regards to `calculateUserAmountOwedForBandwidth`, the reporterId can be *email* or *public key*. It may not be worth it to worry about it now, as we may need a better way to calculate these values, since it will be quite easy to cheat the calculations (though it may be noticeable due to asymmetry of reports). I have some ideas here, as we're currently tracking bandwidth for rate limiting purposes.

File: lib/models/user.js (hunk 5 above)
Comment (barbaraliau) [reply lib/models/user.js]: Yes, these were uncommented

File: lib/models/mirror.js (hunk 2 above)
Comment (barbaraliau) [reply lib/models/mirror.js:63]: Every other model that does a `.save()` follows this pattern. Is there a reason why `mirror.save` wouldn't?

Comment (braydonf) [reply lib/models/mirror.js:63]: Hmm.... I didn't realize that.

Comment (braydonf) [reply lib/models/mirror.js:63]: Actually it looks like this is common:
```js
shard.save(function(err) {
  if (err) {
    return callback(err);
  }
  callback(null, shard);
});
```

Is shard not returned back as the second argument of save? And likewise in this case mirror? If it is, then it can just be, as it was before:
```js
mirror.save(callback);
```

**Comment by braydonf** [comment]
Looks like there is a missing dependency for the tests, and why it's failing.

File: lib/models/mirror.js (hunk 1 above)
Comment (braydonf) [reply lib/models/mirror.js:6]: Part of the reason I had thought to have this be a separate PR,  since I don't think this is related to billing, but if you want to include with this one please just coordinate w/ dylan.

File: lib/models/user.js [hunk 6]
```diff
@@ -6,6 +6,129 @@ const activator = require('hat').rack(256);
...
+PaymentProcessor.virtual('billingDate').get(function() {
+  return this.adapter.billingDate();
+});
+
+PaymentProcessor.virtual('paymentMethods').get(function() {
+  return this.adapter.paymentMethods();
+});
+
+PaymentProcessor.methods.refresh = function() {
+  console.log('USER REFRESHED.');
```
Comment (braydonf) [inline lib/models/user.js:73]: Stray console log?

File: lib/models/user.js [hunk 7]
```diff
@@ -6,6 +6,129 @@ const activator = require('hat').rack(256);
...
+  return UserModel
+    .findOne({ _id: this.__parent._id })
+    .then((user) => {
+      console.log('USER MODEL REFRESH: ', user.defaultPaymentProcessor.data.customer.sources.data);
```
Comment (braydonf) [inline lib/models/user.js:77]: Another

File: lib/models/user.js [hunk 8]
```diff
@@ -6,6 +6,129 @@ const activator = require('hat').rack(256);
...
+      this.__parent.paymentProcessors.splice(
+        0,
+        this.__parent.paymentProcessors.length,
+        user.paymentProcessors
+      )
+      console.log('__PARENT', this.__parent.paymentProcessors);
```
Comment (braydonf) [inline lib/models/user.js:83]: Another

File: lib/models/user.js [hunk 9]
```diff
@@ -294,5 +461,5 @@ User.statics.lookup = function(email, passwd, callback) {
 };
 
 module.exports = function(connection) {
-  return connection.model('User', User);
+  return (UserModel = connection.model('User', User));
```
Comment (braydonf) [inline lib/models/user.js:464]: What's up here?

File: package.json [hunk 10]
```diff
@@ -27,17 +27,19 @@
   },
   "homepage": "https://github.com/Storj/service-storage-models#readme",
   "dependencies": {
+    "dotenv": "^2.0.0",
     "elliptic": "^6.3.2",
+    "graphql": "^0.8.2",
```
Comment (braydonf) [inline package.json:32]: Not needed here.

File: lib/models/payment-processor-adapters/stripe.js [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+      });
+    });
+  },
+  /**
+   * isSubscribed - returns true if the user has a subscription to
+   * the stripe plan with id of `STRIPE_PLAN_ID`
+   * @return {Boolean}
+   */
+  validate: function() {
+    console.log('stripeProcessor validate: ', stripeProcessor);
```
Comment (braydonf) [inline lib/models/payment-processor-adapters/stripe.js:105]: Another stray console log

File: lib/models/payment-processor-adapters/stripe.js [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+    // TODO: verify with stripe that subscription/customer are still valid, etc.
+    const subscriptions = stripeProcessor.data.customer.subscriptions;
+
+    console.log("stripe payment processor info: ", subscriptions);
```
Comment (braydonf) [inline lib/models/payment-processor-adapters/stripe.js:109]: Another

File: lib/models/payment-processor-adapters/stripe.js [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+
+    if (subscriptions.length > 1) {
+      // TODO: add actual error handling...
+      console.log('subscription length > 1');
```
Comment (braydonf) [inline lib/models/payment-processor-adapters/stripe.js:113]: Another

File: lib/models/payment-processor-adapters/stripe.js [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+      return Promise.reject(new Error('customer has more than one stripe subscription!'));
+    }
+
+    if (subscriptions.length < 1) {
+      console.warn('customer has no subscriptions!');
```
Comment (braydonf) [inline lib/models/payment-processor-adapters/stripe.js:118]: Another

File: lib/models/payment-processor-adapters/stripe.js [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+      return Promise.resolve(false);
+    }
+
+    if (subscriptions.data[0].plan.id !== STRIPE_PLAN_ID) {
+      // TODO: add actual error handling...
+      console.log('stripe plan id wrong');
```
Comment (braydonf) [inline lib/models/payment-processor-adapters/stripe.js:124]: Another

File: lib/models/user.js [hunk 16]
```diff
@@ -238,29 +407,27 @@ User.statics.create = function(email, passwd, callback) {
...
-    );
+    return callback(new Error('Password must be hex encoded SHA-256 hash'));
   }
 
-  User.findOne({ _id: email }, function(err, result) {
+  User.findOne({
+    _id: email
+  }, function(err, result) {
     if (err) {
-      return callback(new errors.InternalError(err.message));
```
Comment (braydonf) [inline lib/models/user.js:248]: I think these were already fixed in @barbaraliau pull request, but these should remain as the specific error types.

File: lib/models/user.js [hunk 17]
```diff
@@ -238,29 +407,27 @@ User.statics.create = function(email, passwd, callback) {
   // Check to make sure the password is already SHA-256 (or at least is the
   // correct number of bits).
   if (Buffer(passwd, 'hex').length * 8 !== 256) {
-    return callback(
-      new errors.BadRequestError('Password must be hex encoded SHA-256 hash')
-    );
+    return callback(new Error('Password must be hex encoded SHA-256 hash'));
```
Comment (braydonf) [inline lib/models/user.js:410]: Should remain as specific error type

File: lib/models/user.js [hunk 18]
```diff
@@ -238,29 +407,27 @@ User.statics.create = function(email, passwd, callback) {
...
+    _id: email
+  }, function(err, result) {
     if (err) {
-      return callback(new errors.InternalError(err.message));
+      return callback(err);
     }
 
     if (result) {
-      return callback(new errors.BadRequestError('Email is already registered'));
+      return callback(new Error('Email is already registered'));
```
Comment (braydonf) [inline lib/models/user.js:421]: Should remain as specific error type.

File: lib/models/user.js [hunk 19]
```diff
@@ -238,29 +407,27 @@ User.statics.create = function(email, passwd, callback) {
...
     }
 
     user.save(function(err) {
       if (err) {
         if (err.code === 11000) {
-          return callback(
-            new errors.BadRequestError('Email is already registered')
-          );
+          return callback(new Error('Email is already registered'));
```
Comment (braydonf) [inline lib/models/user.js:427]: Should remain as specific error type

File: lib/models/user.js [hunk 20]
```diff
@@ -238,29 +407,27 @@ User.statics.create = function(email, passwd, callback) {
...
         }
 
-        return callback(new errors.InternalError(err.message));
+        return callback(err);
```
Comment (braydonf) [inline lib/models/user.js:430]: Should remain as specific error type.

File: lib/models/user.js [hunk 21]
```diff
@@ -6,6 +6,129 @@ const activator = require('hat').rack(256);
...
+      )
+      console.log('__PARENT', this.__parent.paymentProcessors);
+      return this.__parent.paymentProcessors;
+    })
+};
+
+PaymentProcessor.methods.update = function(paymentProcessor) {
+  return this.adapter.validate(paymentProcessor)
+    .then((isValid) => {
+      // console.log('this is: ', this);
```
Comment (braydonf) [inline lib/models/user.js:91]: nit: stray commented log

File: lib/models/mirror.js (hunk 2 above)
Comment (barbaraliau) [reply lib/models/mirror.js:63]: Mongoose's .save method returns the err and then the newly saved doc as the second argument. I'm not really sure why it's not returned in the callback.

Comment (braydonf) [reply lib/models/mirror.js:63]: I see, we should be fine to use `mirror.save(callback)` then.

**Comment by barbaraliau** [comment]
Starting new clean pull request.
